### PR TITLE
Doc/move irep docs from util to irep

### PR DESCRIPTION
**Pull request by owen-mc-diffblue**

### CODE REVIEW

File: src/util/irep.h [hunk 1]
```diff
@@ -82,8 +82,77 @@ inline const std::string &name2string(const irep_namet &n)
...
+/// * `comments`:   Another map from `irep_namet` to `irept` which is used
+///   for annotations and other ‘non-semantic’ information
+///
+/// * `sub`:   A vector of `irept` which is used to store ordered but
+///   unnamed children.
+///
+/// The `irept::pretty` function outputs the contents of an `irept` directly
+/// and can be used to understand and debug problems with `irept`s.
+///
+/// On their own `irept`s do not “mean” anything; they are effectively
```
Comment (smowton) [inline src/util/irep.h]: Mix of smart and normal quotes

File: src/util/irep.h [hunk 2]
```diff
@@ -82,8 +82,77 @@ inline const std::string &name2string(const irep_namet &n)
...
+/// things:
+///
+/// * `data`:   A string[^2], which is returned when the `id()` function is
+///   used.
+///
+/// * `named_sub`:   A map from `irep_namet` (a string) to `irept`. This
+///   is used for named children, i.e.  subexpressions, parameters, etc.
+///
+/// * `comments`:   Another map from `irep_namet` to `irept` which is used
+///   for annotations and other ‘non-semantic’ information
```
Comment (smowton) [inline src/util/irep.h]: Probably worth mentioning the caveat about these not being taken into account by the default `operator==` here

File: src/util/irep.h [hunk 3]
```diff
@@ -82,8 +82,77 @@ inline const std::string &name2string(const irep_namet &n)
...
+/// * `sub`:   A vector of `irept` which is used to store ordered but
+///   unnamed children.
+///
+/// The `irept::pretty` function outputs the contents of an `irept` directly
+/// and can be used to understand and debug problems with `irept`s.
+///
+/// On their own `irept`s do not “mean” anything; they are effectively
+/// generic tree nodes. Their interpretation depends on the contents of
+/// result of the `id` function (the `data`) field. `util/irep_ids.txt`
+/// contains the complete list of `id` values. During the build process it
```
Comment (smowton) [inline src/util/irep.h]: Nah, ireps that use other strings (e.g. `irept("Hello world")` are not common but are definitely present in the codebase

File: src/util/irep.h [hunk 4]
```diff
@@ -82,8 +82,77 @@ inline const std::string &name2string(const irep_namet &n)
...
+/// point for working out how to manipulate data.
+///
+/// There are three main groups of classes (or APIs); those derived from
+/// `typet`, `codet` and `exprt` respectively. Although all of these inherit
+/// from `irept`, these are the most abstract level that code should handle
+/// data. If code is manipulating plain `irept`s then something is wrong
+/// with the architecture of the code.
+///
+/// Many of the key descendant of `exprt` are declared in `std_expr.h`. All
+/// expressions have a named subfield / annotation which gives the type of
```
Comment (smowton) [inline src/util/irep.h]: `named subexpression with ID "type"` would probably be clearer

File: src/util/irep.h [hunk 5]
```diff
@@ -82,8 +82,77 @@ inline const std::string &name2string(const irep_namet &n)
...
+/// the expression (slightly simplified from C/C++ as `unsignedbv_typet`,
+/// `signedbv_typet`, `floatbv_typet`, etc.). All type conversions are
+/// explicit with an expression with `id() == ID_typecast` and an ‘interface
```
Comment (smowton) [inline src/util/irep.h]: "interface class" is not used anywhere else, just say "i.e. a `typecast_exprt`" instead

File: src/util/irep.h [hunk 6]
```diff
@@ -82,8 +82,77 @@ inline const std::string &name2string(const irep_namet &n)
...
+/// class’ named `typecast_exprt`. One key descendant of `exprt` is
+/// `symbol_exprt` which creates `irept` instances with the id of “symbol”.
+/// These are used to represent variables; the name of which can be found
+/// using the `get_identifier` accessor function.
+///
+/// `codet` inherits from `exprt` and is defined in `std_code.h`. It
+/// represents executable code; statements in C rather than expressions. In
```
Comment (smowton) [inline src/util/irep.h]: ...a C-like language...

File: src/util/irep.h [hunk 7]
```diff
@@ -82,8 +82,77 @@ inline const std::string &name2string(const irep_namet &n)
...
+/// the front-end there are versions of these that hold whole code blocks,
+/// but in goto-programs these have been flattened so that each `irept`
+/// represents one sequence point (almost one line of code / one
+/// semi-colon). The most common descendants of `codet` are `code_assignt`
```
Comment (smowton) [inline src/util/irep.h]: is `code_assignt`

**Review by smowton — COMMENTED** [review]
Might as well update this while we're doing this

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 105e297b388e75dab6ebfaccbe329aacd0f65f64).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/80008898

File: src/util/dstring.h [hunk 8]
```diff
@@ -16,8 +16,19 @@ Author: Daniel Kroening, kroening@kroening.com
...
-// No virtual destructor, so runtime-polymorphic use would be unsafe.
+/// `dstringt` has one field, an unsigned integer `no` which is an index into
+/// a static table of strings. This makes it expensive to create a new string
+/// (because you have to look through the whole table to see if it is already
+/// there, and add it if it isn't) but very cheap to compare strings (just
+/// compare the two integers). It also means that when you have lots of copies
+/// of the same string you only have to store the whole string once, which
+/// saves space.
+///
+/// `irep_idt` and `irep_namet` are typedef-ed to `dstringt` in irep.h unless
```
Comment (peterschrammel) [inline src/util/dstring.h]: unless what?

File: src/util/irep.h [hunk 9]
```diff
@@ -82,8 +82,77 @@ inline const std::string &name2string(const irep_namet &n)
...
+///
+/// * `sub`:   A vector of `irept` which is used to store ordered but
+///   unnamed children.
+///
+/// The `irept::pretty` function outputs the contents of an `irept` directly
+/// and can be used to understand and debug problems with `irept`s.
+///
+/// On their own `irept`s do not “mean” anything; they are effectively
+/// generic tree nodes. Their interpretation depends on the contents of
+/// result of the `id` function (the `data`) field. `util/irep_ids.txt`
```
Comment (peterschrammel) [inline src/util/irep.h]: This is `irep_ids.def` now.

File: src/util/irep.h [hunk 10]
```diff
@@ -82,8 +82,77 @@ inline const std::string &name2string(const irep_namet &n)
...
+/// * `named_sub`:   A map from `irep_namet` (a string) to `irept`. This
+///   is used for named children, i.e.  subexpressions, parameters, etc.
+///
+/// * `comments`:   Another map from `irep_namet` to `irept` which is used
+///   for annotations and other ‘non-semantic’ information
+///
+/// * `sub`:   A vector of `irept` which is used to store ordered but
+///   unnamed children.
+///
+/// The `irept::pretty` function outputs the contents of an `irept` directly
```
Comment (peterschrammel) [inline src/util/irep.h]: maybe say that it prints the explicit tree structure instead of 'directly'

File: src/util/irep.h [hunk 11]
```diff
@@ -82,8 +82,77 @@ inline const std::string &name2string(const irep_namet &n)
...
+/// CPROVER. `irept` provides a single, unified representation for all of
+/// these, allowing structure sharing and reference counting of data. As
+/// such `irept` is the basic unit of data in CPROVER.  Each `irept`
+/// contains[^1] a basic unit of data (of type `dt`) which contains four
+/// things:
+///
+/// * `data`:   A string[^2], which is returned when the `id()` function is
+///   used.
+///
+/// * `named_sub`:   A map from `irep_namet` (a string) to `irept`. This
```
Comment (peterschrammel) [inline src/util/irep.h]: Use as many \ref as possible

File: src/util/irep.h [hunk 12]
```diff
@@ -82,8 +82,77 @@ inline const std::string &name2string(const irep_namet &n)
...
+/// `codet` inherits from `exprt` and is defined in `std_code.h`. It
+/// represents executable code; statements in C rather than expressions. In
+/// the front-end there are versions of these that hold whole code blocks,
+/// but in goto-programs these have been flattened so that each `irept`
+/// represents one sequence point (almost one line of code / one
+/// semi-colon). The most common descendants of `codet` are `code_assignt`
+/// so a common pattern is to cast the `codet` to an assignment and then
+/// recurse on the expression on either side.
+///
+/// [^1]: Or references, if reference counted data sharing is enabled. It is
```
Comment (peterschrammel) [inline src/util/irep.h]: Maybe inline these footnotes.

File: src/util/irep.h (hunk 11 above)
Comment (owen-mc-diffblue) [reply src/util/irep.h]: Annoyingly, doxygen sometimes can't get the \ref to work if the name is in backticks, but sometimes it can

**Comment by owen-mc-diffblue** [comment]
I have addressed all the review comments

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: ae66e6517acd37db581202a0316d939e960146f8).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/80350969

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 6b0dfb65fb758ba394daa844c58a0b10eae67b9e).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/80355880

**Comment by owen-mc-diffblue** [comment]
@peterschrammel @smowton Please re-review.

File: src/util/irep.h [hunk 13]
```diff
@@ -82,8 +82,78 @@ inline const std::string &name2string(const irep_namet &n)
...
+///   \ref dstringt and thus an integer which is a reference into a string
+///   table.)
+///
+/// * \ref irept::dt::named_sub : A map from `irep_namet` (a string) to \ref
+///   irept. This is used for named children, i.e.  subexpressions, parameters,
+///   etc.
+///
+/// * \ref irept::dt::comments : Another map from `irep_namet` to \ref irept
+///   which is used for annotations and other ‘non-semantic’ information. Note
+///   that this map is ignore by the default \ref operator==.
```
Comment (smowton) [inline src/util/irep.h]: ignore*d*

File: src/util/irep.h [hunk 14]
```diff
@@ -82,8 +82,78 @@ inline const std::string &name2string(const irep_namet &n)
...
+/// The \ref irept::pretty function outputs the explicit tree structure of
+/// an \ref irept and can be used to understand and debug problems with
+/// `irept`s.
+///
+/// On their own `irept`s do not "mean" anything; they are effectively
+/// generic tree nodes. Their interpretation depends on the contents of
+/// result of the \ref id() function, i.e. the `data` field. `util/irep_ids.def`
+/// contains a list of `id` values. During the build process it is used
+/// to generate `util/irep_ids.h` which gives constants for each id
+/// (named `ID_`). You can also make `irep_id`s which do not come from
```
Comment (smowton) [inline src/util/irep.h]: irep_idt?

File: src/util/irep.h [hunk 15]
```diff
@@ -82,8 +82,78 @@ inline const std::string &name2string(const irep_namet &n)
...
+/// inherit from \ref irept, these are the most abstract level that code should
+/// handle data. If code is manipulating plain `irept`s then something is wrong
+/// with the architecture of the code.
+///
+/// Many of the key descendants of \ref exprt are declared in \ref std_expr.h.
+/// All expressions have a named subexpression with ID "type", which gives the
+/// type of the expression (slightly simplified from C/C++ as \ref
+/// unsignedbv_typet, \ref signedbv_typet, \ref floatbv_typet, etc.). All type
+/// conversions are explicit with an expression with `id() == ID_typecast`
+/// and a \ref typecast_exprt. One key descendant of \ref exprt is \ref
```
Comment (smowton) [inline src/util/irep.h]: The typecast_exprt *is* the expression with `id() == ID_typecast`

**Review by smowton — APPROVED** [review]

**Review by peterschrammel — APPROVED** [review]

**Review by allredj — COMMENTED** [review]
This PR failed Diffblue compatibility checks (cbmc commit: 08558726bc3de3d50ca38aba91a204f6d4091a24).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/80546022
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).
##

**Pull request merged**